src/env.h

```diff
@@ -459,6 +459,38 @@ class TickInfo : public MemoryRetainer {
   AliasedUint8Array fields_;
 };
 
+class ExitInfo : public MemoryRetainer {
+ public:
+  inline const AliasedInt32Array& fields();
+  inline bool has_exit_code() const;
+  inline int32_t exit_code() const;
+
+  SET_MEMORY_INFO_NAME(ExitInfo)
+  SET_SELF_SIZE(ExitInfo)
+  void MemoryInfo(MemoryTracker* tracker) const override;
+
+  ExitInfo(const ExitInfo&) = delete;
+  ExitInfo& operator=(const ExitInfo&) = delete;
+  ExitInfo(ExitInfo&&) = delete;
+  ExitInfo& operator=(ExitInfo&&) = delete;
+  ~ExitInfo() = default;
+
+  struct SerializeInfo {
+    AliasedBufferIndex fields;
+  };
+  SerializeInfo Serialize(v8::Local<v8::Context> context,
+                          v8::SnapshotCreator* creator);
+  void Deserialize(v8::Local<v8::Context> context);
+
+  enum Fields { kExitCode = 0, kHasExitCode, kFieldsCount };
+
+ private:
+  friend class Environment;  // So we can call the constructor.
+  explicit ExitInfo(v8::Isolate* isolate, const SerializeInfo* info);
+
+  AliasedInt32Array fields_;
```

Comment:
If this is just one AliasedInt32Array, do we really need a special struct for it? I think it would be simpler to just add that buffer to the Environment, similar to the `exiting` or `stream_base_state`. Or, we might as well just rename the `exiting` buffer to `exit_info` and increase its length by one, and use the second item for exit codes (turning it into `AliasedInt32Array` wouldn't matter because `exiting` was intended for a boolean anyway).

Reply:
Thanks for the review.

> If this is just one AliasedInt32Array, do we really need a special struct for it? 

Once seeing structs like `ImmediateInfo` or `TickInfo`, they also have just one AliasedTypedArray. The criteria seem to conflict with existing structs.

I think it's better to encapsulate data that are an AliasedInt32Array and its index identifiers (`enum { kExisting = 0, kExitCode, kHasExitCode }`). And once seeing other structs, I feel creating a new struct similar to them is required for consistency.

The AliasedTypedArray variables in the Environment seem reasonable for the following reason:

- `exiting_` and `should_abort_on_uncaught_toggle_` have one length. So it's clear where its value should be set.
- `stream_base_state_` has a length greater than one but there is a related class where we can bundle index identifiers.

Adding such a struct to the Environment currently leads to changes required here and there. One concern I have is that such changes might be a burden to maintain the snapshot. So it would be better if we can use a struct encapsulating the exit data but not need to de/serialize it.

Reply:
> One concern I have is that such changes might be a burden to maintain the snapshot. 

I think it would be simpler to just reuse the existing buffer? The extra burden is already added in this PR with this special struct. By reusing the buffer there's no need to implement the de/serialization of that buffer. Increasing the length of it doesn't entail further changes, you can just add some getters/setters for the second item and that's it?

Reply:
> Once seeing structs like ImmediateInfo or TickInfo, they also have just one AliasedTypedArray. The criteria seem to conflict with existing structs.

Right, these structs predated MemoryRetainer or the snapshot support. If they were invented now, I'd still prefer to have them inlined to avoid the extra complexity of implementing additional methods..

Reply:
> structs like AsyncHooks are complicated enough to have its own struct, but creating structs for these simpler data just results in more code overall compared to having them inlined

Totally agreed with your thoughts. I will change the PR. Thanks for the clarification.